nitrocli/src/options.rs

```diff
@@ -0,0 +1,155 @@
+
+// *************************************************************************
+// * Copyright (C) 2017-2018 Daniel Mueller (deso@posteo.net)              *
+// *                                                                       *
+// * This program is free software: you can redistribute it and/or modify  *
+// * it under the terms of the GNU General Public License as published by  *
+// * the Free Software Foundation, either version 3 of the License, or     *
+// * (at your option) any later version.                                   *
+// *                                                                       *
+// * This program is distributed in the hope that it will be useful,       *
+// * but WITHOUT ANY WARRANTY; without even the implied warranty of        *
+// * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the         *
+// * GNU General Public License for more details.                          *
+// *                                                                       *
+// * You should have received a copy of the GNU General Public License     *
+// * along with this program.  If not, see <http://www.gnu.org/licenses/>. *
+// *************************************************************************
+
+use std::fmt;
+use std::io;
+use std::str::FromStr;
+
+use crate::commands;
+use crate::error::Error;
+use crate::Result;
+
+/// A top-level command for nitrocli.
+#[derive(Debug)]
+pub enum Command {
+  Clear,
+  Close,
+  Open,
+  Status,
+}
+
+impl Command {
+  /// Execute this command with the given arguments.
+  pub fn execute(&self, args: Vec<String>) -> Result<()> {
+    match *self {
+      Command::Clear => clear(args),
+      Command::Close => close(args),
+      Command::Open => open(args),
+      Command::Status => status(args),
+    }
+  }
+}
+
+impl fmt::Display for Command {
+  fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
+    write!(
+      f,
+      "{}",
+      match *self {
+        Command::Clear => "clear",
+        Command::Close => "close",
+        Command::Open => "open",
+        Command::Status => "status",
+      }
+    )
+  }
+}
+
+impl FromStr for Command {
+  type Err = ();
+
+  fn from_str(s: &str) -> std::result::Result<Self, Self::Err> {
+    match s {
+      "clear" => Ok(Command::Clear),
+      "close" => Ok(Command::Close),
+      "open" => Ok(Command::Open),
+      "status" => Ok(Command::Status),
+      _ => Err(()),
+    }
+  }
+}
+
+/// Executes the given parser on the given arguments and handles the result.
+///
+/// This macro executed the given argument parser on the given arguments using the standard output
+/// and error.  It then handles three cases:  If the result is `Ok`, it does nothing as the
+/// arguments have been parsed successfully.  If the result is `Err(0)`, it returns `Ok(())` as the
+/// help message has been printed by `argparse`.  Otherwise it returns `Err(Error::ArgparseError)`
+/// as `argparse` failed.
+macro_rules! parse_args {
```

Comment:
I guess I do understand why you have introduced a macro here, but I think that can be solved with a function and in the process be made more clear. Would it work to simply propagate the error reported by `argparse` up (i.e., extend `Error::ArgparseError` with the error code) and then make the decision as to how to exit in the `run` function?

Largely untested: [argparse.txt](https://github.com/d-e-s-o/nitrocli/files/2704772/argparse.txt)